### PR TITLE
docs: add callbacks documentation (#564)

### PR DESCRIPTION
**Pull request by LMMilliken**
Added page for callbacks in the documentation, still need to add descriptions for BestModelCheckpoint and TrainingCheckpoint

---

- [x] This PR references an open issue
- [x] I have added a line about this change to CHANGELOG

### CODE REVIEW

File: docs/walkthrough/using-callbacks.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+(using-callbacks)=
+
+# Using Callbacks
+
+Callbacks are a way of adding additional methods to the finetuning process. The methods are executed when certain events occur and there are several callback classes, each serving a different function by providing different methods for different events.
+A run can be assigned multiple callbacks using the optional `callbacks` parameter when it is created.
+
```
Comment (bwanglzu) [inline docs/walkthrough/using-callbacks.md:7]: add some code blocks to demonstrate how to add callbacks, for exmaple:

```python
import finetuner
from finetuner.callbacks import Callabck1, Callback2
finetuner.fit(
    callbacks=[Callback1(), Callback2(),..]
)
```

File: docs/walkthrough/using-callbacks.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+## Evaluation
+
+The evaluation callback is used to calculate performance metrics for the model being tuned at the end of each epoch. In order to evaluate the model, two additional data sets - a query dataset and an index dataset - need to be provided as arguments. If no index set is provided, the query dataset is reused instead. Below is an example of the metrics as they are output at the end of finetuning:
```
Comment (bwanglzu) [inline docs/walkthrough/using-callbacks.md:10]: try to explain what is the difference between `eval_data` and `EvaluationCallback`:

`eval_data` is a DocumentArray, used to evaluate machine learning losses, e.g. if you are fine-tuning CLIP, it's mearesuing CLIPLoss.

`EvaluationCallback` is evaluating the "search metrics", such as `average precision` ...

```python
import finetuner

run = finetuner.fit(
    model='some model',
    train_data=MyDocumentArray(..),
    eval_data=MyEvaluatioDocumentArray(...), # eval ML losses
    callbacks=[EvaluationCallback(..)], # evaluating how model is performed on information retrieval metrics.
)
```

File: docs/walkthrough/using-callbacks.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+## Early Stopping
+
+The early stopping callback measures a given metric at the end of every epoch, and will stop the fitting early if the metric doesnt increase at a high enough rate. The best performing model out of the completed epochs is stored in the case that it starts to perform worse after a certain point. Both the minimum improvement and the number of successive rounds without improvement before the training ends can be specified as optional arguments at construction.
+The example output below shows a situation in which the early stopping callback is used to end a training run early. Compared to the results above in the evaluation explanation, the number of epochs was only two, compared to the full five, more than halfing the runtime of the training. On the other hand, the final loss recorded is 0.007, more than twice as high as if all five training epochs had completed. While the parameters of the early stopping callback can be adjusted to further improve either the runtime or loss, however this tradeoff between the two outcomes is what the early stopping callback promises to deliver.
+
```
Comment (bwanglzu) [inline docs/walkthrough/using-callbacks.md:35]: add some code block here demonstrate the parameters and default ones and how to use them

**Review by bwanglzu — COMMENTED** [review]
great starting point! left some general comments

File: docs/walkthrough/using-callbacks.md (hunk 1 above)
Comment (LMMilliken) [reply docs/walkthrough/using-callbacks.md]: I have added some example code now

File: docs/walkthrough/using-callbacks.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+           DEBUG    Metric: 'resnet50_reciprocal_rank' Value: 0.16654                                    __main__.py:205
+           INFO     Building the artifact ...                                                            __main__.py:207
+           INFO     Pushing artifact to Hubble ...                                                       __main__.py:231
+```
+
+The evaluation callback is triggered at the end of each epoch, in which the model is evaluated using the `query_data` and `index_data` datasets that were provided when the callback was created. It is worth noting that the evaluation callback and the eval_data parameter of the fit method do not do the same thing. The eval data parameter takes a Document Array (or the name of one that has been pushed on Hubble) and uses its contents to evaluate the loss of the model whereas the evaluation callback is used to evaluate the quality of the searches using metrics such as average precision and recall. These search metrics can be used by other callbacks if the evaluation callback is first in the list of callbacks when creating a run.
```
Comment (bwanglzu) [inline docs/walkthrough/using-callbacks.md:45]: Document Array -> `DocumentArray`

File: docs/walkthrough/using-callbacks.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        EvaluationCallback(
+            query_data='tll-test-query-da',
+            index_data='tll-test-index-da'
+        ),
+        EarlyStopping()
+    ]
+)
+```
+
+## Evaluation
```
Comment (bwanglzu) [inline docs/walkthrough/using-callbacks.md:24]: EvaluationCallback

File: docs/walkthrough/using-callbacks.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                    __main__.py:196
+           INFO     Building the artifact ...                                        
+                    __main__.py:207
+           INFO     Pushing artifact to Hubble ...                                   
+                    __main__.py:231
+```
+
+This callback is triggered at the end of both training and evaluation batches, to record the losses of the two data sets, and is triggered a third time at the end of each epoch to evaluate the performance of the current model using the monitored metric and then record this model if it performs better than the best model so far.
+
+## Early Stopping
```
Comment (bwanglzu) [inline docs/walkthrough/using-callbacks.md:76]: EarlyStopping

**Review by bwanglzu — CHANGES_REQUESTED** [review]
left some minor comments, can we break down long paragraphs into shorter sentences? or listed items to be more direct and infomative?

**Comment by LMMilliken** [comment]
> left some minor comments, can we break down long paragraphs into shorter sentences? or listed items to be more direct and infomative?

I have restructured several parts, should be nicer to read now

File: docs/walkthrough/using-callbacks.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+## BestModelCheckpoint
+
+This callback evaluates the performance of the model at the end of each epoch, and keeps a record of the best perfoming model across all epochs. Once fitting is finished the best performing model is saved instead of the most recent model. The definition of best is based on two parameters:
+
+- `monitor`: The metric that is used to compare models to each other. By default this value is 'val_loss', the loss function calculated using the evaluation data, however the loss calculated on the training data can be used instead with 'train_loss'; any metric that is recorded by the evaluation callback can also be used.
+- `mode`: Whether the monitored metric should be maximised ('max') or minimised ('min'). By default the mode is set to auto, meaning that it will automatically chose the correct mode depending on the chosen metric: 'min' if the metric is loss and 'max' if the metric is one recorded by the evaluation callback.
```
Comment (gmastrapas) [inline docs/walkthrough/using-callbacks.md:53]: Same here max and min in code quotes

File: docs/walkthrough/using-callbacks.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                    __main__.py:194
+           DEBUG    Finetuning took 0 days, 0 hours 1 minutes and 16 seconds         
+                    __main__.py:196
+           INFO     Building the artifact ...                                        
+                    __main__.py:207
+           INFO     Pushing artifact to Hubble ...                                   
+                    __main__.py:231
+```
+
+This callback is triggered at the end of both training and evaluation batches, to record the losses of the two data sets, and is triggered a third time at the end of each epoch to evaluate the performance of the current model using the monitored metric and then record this model if it performs better than the best model so far.
```
Comment (guenthermi) [inline docs/walkthrough/using-callbacks.md:79]: I don't know if it is relevant to the user to know what exactly each event causes in the callback. I think the information which is important for the user is already mentioned above. 

If we want to keep this sentence, I would suggest to split it into multiple shorter sentences. Moreover, it should be a little bit more precise, i.e., a third time is technically not correct since it is called before after each batches, and, at the end of the fit process it is also called another time to load the best model.

**Review by gmastrapas — COMMENTED** [review]
Really good use of English 🔥 Left some comments

**Review by guenthermi — CHANGES_REQUESTED** [review]
Looks already quite good. Added some comments

File: docs/walkthrough/using-callbacks.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+## EarlyStopping
+
+Similarly to the best model checkpoint callback, the early stopping callback measures a given metric at the end of every epoch and saves the best performing model at the end of the fitting process. Unlike the best model checkpoint callback, the early stopping callback will end the fitting process early if the metric does not improve enough between successive runs. Since the best model is only used to assess the rate of improvement, only the monitored metric is needed and so the model itself is not saved.
```
[the remark this answers is not in the document]
Comment (LMMilliken) [reply docs/walkthrough/using-callbacks.md]: Corrected now

File: docs/walkthrough/using-callbacks.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            mode = "min",
+            patience=2,
+            min_delta=1,
+            baseline=1.5
+        )
+    ]
+)
+```
+
+The early stopping callback triggers at the same times as the best model checkpoint callback: at the end of training and evaluation batches to record the loss, and at the end of each epoch to evaluate the model and compare it to the best so far. Whether it stops training at the end of an epoch depends on several parameters:
```
[the remark this answers is not in the document]
Comment (LMMilliken) [reply docs/walkthrough/using-callbacks.md]: I have clarified this sentence

File: docs/walkthrough/using-callbacks.md (hunk 8 above)
Comment (LMMilliken) [reply docs/walkthrough/using-callbacks.md]: I agree and have removed this part now

File: CHANGELOG.md [hunk 11]
```diff
@@ -10,6 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Add documentation for callbacks ([#564](https://github.com/jina-ai/finetuner/pull/567))
```
Comment (bwanglzu) [inline CHANGELOG.md:13]: let's always link to the PR 567

File: docs/walkthrough/using-callbacks.md [hunk 12]
```diff
@@ -0,0 +1,126 @@
+(using-callbacks)=
+# Using Callbacks
+
+Callbacks are a way of adding additional methods to the finetuning process. The methods are executed when certain events occur and there are several callback classes, each serving a different function by providing different methods for different events.
+A run can be assigned multiple callbacks using the optional `callbacks` parameter when it is created.
+
+```python
+run = finetuner.fit(
```
Comment (bwanglzu) [inline docs/walkthrough/using-callbacks.md:8]: add a line above to import the callbacks, from finetuner.callback import ...

Comment (LMMilliken) [reply docs/walkthrough/using-callbacks.md:8]: added now

**Review by bwanglzu — APPROVED** [review]
LGTM!

File: docs/walkthrough/using-callbacks.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+           DEBUG    Metric: 'resnet50_reciprocal_rank' Value: 0.16654                                    __main__.py:205
+           INFO     Building the artifact ...                                                            __main__.py:207
+           INFO     Pushing artifact to Hubble ...                                                       __main__.py:231
+```
+
+The evaluation callback is triggered at the end of each epoch, in which the model is evaluated using the `query_data` and `index_data` datasets that were provided when the callback was created.
+It is worth noting that the evaluation callback and the `eval_data` parameter of the fit method do not do the same thing. The eval data parameter takes a `DocumentArray` (or the name of one that has been pushed on Hubble) and uses its contents to evaluate the loss of the model whereas the evaluation callback is used to evaluate the quality of the searches using metrics such as average precision and recall. These search metrics can be used by other callbacks if the evaluation callback is first in the list of callbacks when creating a run.
```
Comment (gmastrapas) [inline docs/walkthrough/using-callbacks.md:46]: let's not mention Hubble, users dont know what Hubble is. You can write cloud instead

**Review by gmastrapas — APPROVED** [review]

File: docs/walkthrough/using-callbacks.md (hunk 13 above)
Comment (LMMilliken) [reply docs/walkthrough/using-callbacks.md]: I've changed this now

**Review by guenthermi — APPROVED** [review]
LGTM
